packages/medusa-payment-paypal/src/services/paypal-provider.ts

```diff
@@ -0,0 +1,386 @@
+import PayPal from "@paypal/checkout-server-sdk"
+import { EOL } from "os"
+import {
+  AbstractPaymentProcessor,
+  isPaymentProcessorError,
+  PaymentProcessorContext,
+  PaymentProcessorError,
+  PaymentProcessorSessionResponse,
+  PaymentSessionStatus,
+} from "@medusajs/medusa"
+import {
+  PaypalOptions,
+  PaypalOrder,
+  PaypalOrderStatus,
+  PurchaseUnits,
+} from "../types"
+import { humanizeAmount } from "medusa-core-utils"
+import { roundToTwo } from "./utils/utils"
+
+class PayPalProviderService extends AbstractPaymentProcessor {
+  static identifier = "paypal"
+
+  protected readonly options_: PaypalOptions
+  protected paypal_: PayPal
+
+  constructor(_, options) {
+    // @ts-ignore
+    super(_, options)
+
+    this.options_ = options
+    this.init()
+  }
+
+  protected init(): void {
+    let environment
+    if (this.options_.sandbox) {
+      environment = new PayPal.core.SandboxEnvironment(
+        this.options_.client_id,
+        this.options_.client_secret
+      )
+    } else {
+      environment = new PayPal.core.LiveEnvironment(
+        this.options_.client_id,
+        this.options_.client_secret
+      )
+    }
+
+    this.paypal_ = new PayPal.core.PayPalHttpClient(environment)
+  }
+
+  async getPaymentStatus(
+    paymentSessionData: Record<string, unknown>
+  ): Promise<PaymentSessionStatus> {
+    const order = (await this.retrievePayment(
+      paymentSessionData
+    )) as PaypalOrder
+
+    switch (order.status) {
+      case PaypalOrderStatus.CREATED:
+        return PaymentSessionStatus.PENDING
+      case PaypalOrderStatus.SAVED:
+      case PaypalOrderStatus.APPROVED:
+      case PaypalOrderStatus.PAYER_ACTION_REQUIRED:
+        return PaymentSessionStatus.REQUIRES_MORE
+      case PaypalOrderStatus.VOIDED:
+        return PaymentSessionStatus.CANCELED
+      case PaypalOrderStatus.COMPLETED:
+        return PaymentSessionStatus.AUTHORIZED
+      default:
+        return PaymentSessionStatus.PENDING
+    }
+  }
+
+  async initiatePayment(
+    context: PaymentProcessorContext
+  ): Promise<PaymentProcessorError | PaymentProcessorSessionResponse> {
+    const { currency_code, amount, resource_id } = context
+
+    let session_data
+
+    try {
+      const request = new PayPal.orders.OrdersCreateRequest()
+      request.requestBody({
```

Comment:
thought(non-blocking): I guess it wouldn't hurt to make the request settings configurable - at least the `intent` method.

Reply:
Yes, I think a global plugin option makes sense. Would make it consistent with the Stripe plugin 👍